src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc

```diff
@@ -479,63 +600,122 @@ void Codegen::bundleFinish(
-// Does not deal with the control mode and digital interface of the instrument.
-void Codegen::customGate(
-    const Str &iname,
-    const Vec<UInt> &operands,
-    const Vec<UInt> &creg_operands,
-    const Vec<UInt> &breg_operands,
-    ir::compat::ConditionType condition,
-    const Vec<UInt> &cond_operands,
-    Real angle,
-    UInt startCycle, UInt durationInCycles
-) {
-#if 0   // FIXME: test for angle parameter
-    if(angle != 0.0) {
-        QL_DOUT("iname=" << iname << ", angle=" << angle);
+void Codegen::custom_instruction(const ir::CustomInstruction &custom) {
+    Operands ops;
+
+    // Handle the template operands for the instruction_type we got. Note that these are empty if that is a 'root'
+    // InstructionType, and only contains data if it is one of the specializations (see ir.gen.h)
+    // FIXME: check for existing decompositions (which should have been performed already by an upstream pass)
+    /*
+     FIXME: these are operands that match a specialized instruction definition, e.g. "cz q0,q10"
+     FIXME: these are not handled below, so things fail if we have such definitions
+    cQASM "cz q[0],q[10]" with JSON "cz q0,q10" results in:
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/backend.cc:152 custom instruction: name=cz
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc:814 found template_operands: JSON = {"cc":{"signal":[],"static_codeword_override":[0]},"duration":40}
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc:798 template operand: q[0]
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc:798 template operand: q[10]
+
+    cQASM "cz q[0],q[10]" with JSON "cz q0,q9" results in:
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/backend.cc:152 custom instruction: name=cz
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc:814 found template_operands: JSON = {"cc":{"signal":[],"static_codeword_override":[0]},"duration":40}
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc:798 template operand: q[0]
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc:809 operand: q[10]
+
+    cQASM "cz q[0],q[10]" with JSON "cz q1,q10" results in:
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/backend.cc:152 custom instruction: name=cz
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc:829 operand: q[0]
+    [OPENQL] /Volumes/Data/shared/GIT/OpenQL/src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc:829 operand: q[10]
+    */
+#if 1   // FIXME
+    if(!custom.instruction_type->template_operands.empty()) {
+        QL_DOUT("found template_operands: JSON = " << custom.instruction_type->data.data );
```

Comment:
The texts above with [OPENQL] were not clear to me. I agree that you insist on getting specialized instructions. Then that decomposition should have been done before codegen. Is that the message of those texts?

Reply:
The texts with "[OPENQL]" are saved log messages for different combinations of operands being template operands or plain operands. For now, the existence of template operands is forbidden (in the backend), which means that gate decomposition must have been performed successfully.
This is still there because it does need work, especially to behave well on misconfiguration.